### ISSUE TITLE
Running grpc server on Android Platform

### ISSUE THREAD

**Issue by zothen** [grpc/grpc-java#3356]
Please answer these questions before submitting your issue.

### What version of gRPC are you using?
1.5

### What JVM are you using (`java -version`)?
1.7

### What did you do?
If possible, provide a recipe for reproducing the error.

**Goal**

Run the HelloWorldServer grpc-java example in a system app running on a Marshmallow Android platform.

**Steps**

1. Follow the grpc-java documentation for retrieving and buiding the examples.  The build environment is created by sourcing the AOSP file build/envsetup.sh, then running the lunch command for our platform.  I'm guessing that you could use an emulated lunch platform or a Nexus/Pixel lunch platorm to recreate this step.  

2.  Use the jar files built from step 1 and necessary HelloWorld java file in an AOSP system app project which will create the HelloWorld grpc server just like the grpc-java example.  In the Android.mk file, proguard is disabled so nothing should be removed.  Here is the list of jars from step 1 that are included in the project:

api-common-1.0.0-rc1.jar
auto-value-1.1.jar
error_prone_annotations-2.0.19.jar
examples.jar
grpc-context-1.5.0.jar
grpc-core-1.5.0.jar
grpc-netty-1.5.0.jar
grpc-protobuf-1.5.0.jar
grpc-protobuf-lite-1.5.0.jar
grpc-stub-1.5.0.jar
gson-2.7.jar
guava-19.0.jar
instrumentation-api-0.4.3.jar
joda-time-2.8.2.jar
jsr305-3.0.0.jar
netty-buffer-4.1.12.Final.jar
netty-codec-4.1.12.Final.jar
netty-codec-http2-4.1.12.Final.jar
netty-codec-http-4.1.12.Final.jar
netty-codec-socks-4.1.12.Final.jar
netty-common-4.1.12.Final.jar
netty-handler-4.1.12.Final.jar
netty-handler-proxy-4.1.12.Final.jar
netty-resolver-4.1.12.Final.jar
netty-transport-4.1.12.Final.jar
protobuf-java-3.3.1.jar
protobuf-java-util-3.3.1.jar
proto-google-common-protos-0.1.9.jar

More details can be provided as needed.

### What did you expect to see?

I expected the HelloWorld server to properly start.

### What did you see instead?

When the system app executes, I see the following message in logcat:

io.grpc.ManagedChannelProvider$ProviderNotFoundException: No functional server found. Try adding a dependency on the grpc-netty artifact

Note: This issue is similar to the close issue #2227  but in my case, I'm trying to build and run a system app on an Android Platform which I control/own.

**Comment by ejona86** [comment]
We don't support Netty (client or server) on Android. However, I have seen it work. The `ProviderNotFoundException` can be fixed with ProGuard configuration of:
```
-keep io.grpc.netty.NettyServerProvider
```

**Comment by zothen** [comment]
Hey Eric, thanks for the quick response.

Per my reproduction steps, I have proguard completely disabled.  If there is any other information that I can gather to help debug this, just let me know.

Thanks,
Rob

**Comment by ejona86** [comment]
I have no knowledge of the `Android.mk` build. If it's not proguard, then I'd assume the `META-INF/services/` files were being stripped. That's not a problem when building Android Apps using Gradle, but maybe it is when using that other build system. You can also double-check that `io/grpc/netty/NettyServerProvider` (.class/.dex) is present in the resulting build.

**Comment by zothen** [comment]
On an AOSP Android.mk build of java code, disabling proguard results in zero stripping.  I've decompiled the apk I created and verified that the io/grpc/netty/NettyServerProvider class is present.

In my reproduction steps, I've listed all the grpc-java jars and given that proguard/stripping is disabled, so with stripping disabled, all those classes should be present.  Perhaps my method of building the grpc-java examples, then including those jars in my project is flawed.  Perhaps additional jars are needed that exist on my Ubuntu workstation by not on the Android platform?

Note that unlike the issue of trying get grpc-java running in a app that needs to run on any Android device, in my case I own the platform so I can add system priveleged libraries as needed.

Another bit of info worth highlighting is the the Android Marshmallow toolchain that I'm using to build the grpc-java examples uses JDK 7 in case that is a problem.  I realized that the TLS functionality may require JDK 8 or other changes, but for now, I was just trying to get the HelloWorldServer example to work.

Thanks,
Rob

**Comment by ejona86** [comment]
We don't need JDK 8 for Android, since Android uses Conscrypt instead of the normal TLS implementation.

Since `NettyServerProvider` is present, that means [`META-INF/services/io.grpc.ServerProvider`](https://github.com/grpc/grpc-java/blob/v1.5.0/netty/src/main/resources/META-INF/services/io.grpc.ServerProvider) must be missing.

**Comment by ejona86** [comment]
Actually, there is one other possibility: the context class loader isn't finding the ServerProvider because it is the wrong ClassLoader. In 1.6 we'll no longer be using the context class loader by default. The only time I've seen it be a problem on android is when you are running multiple apps in the same process.

**Comment by zothen** [comment]
With your help (and after much pain and headbanging) I think with your help I am past this issue and on to the next.

In the Marshmallow AOSP build system, the java build system, Jack, does not create a services directory inside the META-INF when the jars and apks are created, so that was the problem.  Figuring how to make it does was painful as all Googling results pointed to directions for Gradle or Maven regarding the META-INF/services issue.

To make this work in an Android.mk build system, I did the following:

1.  unzip '*.jar" on a directory containing all the grpc jar files.  This created an aggregate META-INF/service directory with the following files:

io.grpc.ManagedChannelProvider
io.grpc.NameResolverProvider
io.grpc.ServerProvider
javax.annotation.processing.Processor

2.  In my apk project directory, I created a directory call meta-inf-adds.  In this directory I created the META-INF/services directory, exactly how it should appear in the apk file.  I then copied the above files to this services directory.

3.  I added the line LOCAL_JACK_FLAGS := $(LOCAL_PATH)/meta-inf-adds to my Android.mk for the apk target.

Those steps did the trick and now I seem to get past the original problem and am on to the next.  I had also tried using LOCAL_JAVA_RESOURCE_DIRS but it seems to only take files from one dir and place them in the root directory of the apk, but the ServiceLoader (or whatever needs to find NettyServerProvider, etc.) isn't looking in that location.

Now I properly get my "Server started, listening on 50051" debug message.  I'm not having some issue when attempting to connect with a grpc client which connects fine if I run grpc-java from an Ubuntu machine, so I'm on to debugging that issue.

Thanks Eric for your help so far!
